### DISCUSSION TITLE
Fedimint Zaplockers

### DISCUSSION THREAD

**Discussion by Kodylow** [fedimint/fedimint#3486] [Fedimint Modules]
In light of Evan's **Attack On The Lightning Network** (TM) announced last week: https://blog.zeusln.com/zeus-v0-8-0-open-beta/ , I started thinking about how we might be able to accomplish something similar with Fedimint: async lightning receives either through a lightning address scheme or something similar.

### How Zaplocker works (copied from zaplocker code):

See zaplocker readme for full description, condensed here with diagram:

Basic Protocol:
1. User Registration: The user logs in, selects a username, and shares payment hashes, signatures, and nostr details with Zaplocker. Zaplocker provides the user with their unique lightning address and a dashboard to view pending payments.

2. Invoice Request: When a sender wants to make a payment, they request an invoice from Zaplocker. Zaplocker creates a hodl invoice using one of the user's payment hashes and notifies the user of the pending payment.

3. User Actions: The user logs in to view and manage their pending payments.

4. Settle on Lightning: To settle the payment on lightning, the user provides Zaplocker with a custom lightning invoice that has a matching payment hash. Zaplocker validates the invoice, pays it, and uses the proof of payment to settle the sender's payment.

5. Update Payment Hashes (Optional): The user can provide new payment hashes and signatures to Zaplocker to update their account, ensuring they always have payment hashes available for transactions.

```mermaid
sequenceDiagram
    participant User
    participant Zaplocker
    participant Sender

    Note over User,Zaplocker: User Registration
    User->>Zaplocker: Provide payment hashes, signatures, nostr details, & choose username
    Zaplocker->>User: Display lightning address & dashboard

    Note over User,Zaplocker: Invoice Request
    Sender->>Zaplocker: Request invoice for payment to User
    Zaplocker->>Sender: Create & provide hodl invoice using User's payment hash
    Sender->>Zaplocker: Initiate Hodl Invoice payment (stays pending)
    Zaplocker->>User: Notify of pending payment via nostr

    Note over User,Zaplocker: User Actions
    User->>Zaplocker: Log in & view pending payment
    User->>User: Decide to settle on lightning

    Note over User,Zaplocker: Settle on Lightning
    User->>Zaplocker: Provide custom lightning invoice with matching payment hash
    Zaplocker->>User: Validate invoice & pay it
    Zaplocker->>Sender: Receive proof of payment & settle sender's payment

    Note over User,Zaplocker: Update Payment Hashes (Optional)
    User->>Zaplocker: Provide new payment hash & commit signature
    Zaplocker->>User: Update user's payment hashes
```

The major issue here is that the hodl invoice locks up liquidity throughout the network on the lightning side. I believe that we can have the liquidity lockup occur fedimint side via the lightning gateway locking up the ecash to a nostr pubkey which stays unclaimed fedimint side until the user logs back on and sweeps the ecash. This addresses the lightning issues and i believe is a unique offering we can use fedimint for because we're running the internal consensus.

Haven't thought this all the way through but think it would require the nostr module first so users could register a persistent pubkey with the federation guardians, and the guardians would be building the lightning side transaction for the user instead of initiating by the client. Let me know your guys' thoughts

**Comment by okjodom** [comment]
Discussion seems like an evolution of #1516

**Comment by m1sterc001guy** [comment]
Ok correct me if I'm missing something here, but since receiving into a Fedimint is already asynchronous, Fedimint should already support a simpler version of this. I think the receiving client could just create/upload a bunch of invoices and their nostr npub/NIP-05 and then send those to the federation (or zaplocker) to hold onto. When a payment is made, the federation could use the npub to notify the receiving client to come back online and claim the ecash (i.e call `await-invoice`).  

So no need for hodl invoices and the bottom half of the graph that settles via lightning. We could potentially even just build this into the existing lightning module.

**Reply by Kodylow** [reply]
Yeah that's what I was thinking, the async part occurs fedimint side with no hodl invoices.

To clarify, what would a client need to provide to the lightning gateway/guardians: you said upload invoices but I don't think they'd even need to do that, could they just register their nostr npub and the gateway completes the lightning side payment by proving they locked ecash to the recipient npub? then guardians notify recipient via nostr message?

What would the client need to update at that point? I think the guardian would have to create the invoice and reveal the preimage right? you couldn't trust the gateway to receive the payment solo it would have to complete the payment by providing the proof to the guardians so the guardians would have to build the transaction and provide to the gateway?

**Comment by Kodylow** [comment]
Was thinking about this again in light of the ocean mining custodianship until payouts kerfuffle .

My understanding is that currently all you need to start mining with them is to point your hash there and give a Bitcoin address, they associate payouts for  your shares with that address and accumulate until you hit 1mil satoshis, at which point they do an onchain payout. This has several weaknesses:

1. Privaxy bc of Address reuse. There is an onchain tie between all of your payouts, and no way I see them being able to improve this onchain excveptr maybe having it be a paynym they submit as the identifier for their hash instead of an address.
2. Custodianship of funds up until payout. The idea is for this to be plebminers, but mining 1mil coins takes some significant time or capex.

I think doing a fedimint payout model like the one we describe here for zaplockers would be ideal.

1. Hasher registers pubkey, nostr or ephemeral, with pool.
2. Mine directly ryo pegin address. Great for fedimint already bc of railing chain tip, 
3. Payout async via locking ecash against the registered pubkey.
4. Users redeem locked ecash to claim, accumulate until they can do a pegout. Larger hashers or hashers with lightning nodes act as lightning gateway.

I think this kills a couple birds with 1 stone. First , fedimint already operates on this "only do large onchain tx for pegin or pegout" model miners have to follow. Ocean and who ever they split guardian running wouldnt have to run lightning at all, federation members running nodes would. Second, our model where the ecash is against onchain Bitcoin instead of against lightning like on Cashu is a huge benefit here. The mining pool doesn't touch lightning at all.
